Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        m_buffer[m_length + 2] = 'u';
+        m_buffer[m_length + 3] = 'e';
+        m_length += 4;
+        return;
+    }
+
+    if (value.isFalse()) {
+        if (remainingCapacity() < 5) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = 'f';
+        m_buffer[m_length + 1] = 'a';
+        m_buffer[m_length + 2] = 'l';
+        m_buffer[m_length + 3] = 's';
+        m_buffer[m_length + 4] = 'e';
+        m_length += 5;
+        return;
+    }
+
+    if (value.isInt32()) {
+        auto number = value.asInt32();
+        auto length = lengthOfIntegerAsString(number);
+        if (remainingCapacity() < length) {
+            recordBufferFull();
+            return;
+        }
+        writeIntegerToBuffer(number, &m_buffer[m_length]);
+        m_length += length;
+        return;
+    }
+
+    if (!value.isCell()) {
+        recordFailure("value type"_s);
+        return;
+    }
+    auto& cell = *value.asCell();
+
+    switch (cell.type()) {
+    case StringType: {
+        auto& string = asString(&cell)->tryGetValue();
+        if (string.isNull()) {
+            recordFailure("String::tryGetValue"_s);
+            return;
+        }
+        if (!string.is8Bit()) {
+            recordFailure("16-bit string"_s);
+            return;
+        }
+        auto stringLength = string.length();
+        if (remainingCapacity() < 1 + stringLength + 1) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = '"';
+        auto* characters = string.characters8();
+        for (unsigned i = 0; i < stringLength; ++i) {
+            auto character = characters[i];
+            if (WTF::escapedFormsForJSON[character]) {
+                recordFailure("string character needs escaping"_s);
+                return;
+            }
+            m_buffer[m_length + 1 + i] = character;
+        }
+        m_buffer[m_length + 1 + stringLength] = '"';
+        m_length += 1 + stringLength + 1;
+        return;
+    }
+
+    case ObjectType:
+    case FinalObjectType: {
+        auto& object = *asObject(&cell);
+        if (object.isCallable()) {
+            recordFailure("callable object"_s);
+            return;
+        }
+        auto* ancestorClass = object.classInfo();
+        do {
+            if (ancestorClass == NumberObject::info()) {
+                recordFailure("wrapped number"_s);
+                return;
+            }
+            if (ancestorClass == StringObject::info()) {
+                recordFailure("wrapped string"_s);
+                return;
+            }
+            if (ancestorClass == BooleanObject::info()) {
+                recordFailure("wrapped boolean"_s);
+                return;
+            }
+            if (ancestorClass == BigIntObject::info()) {
+                recordFailure("wrapped big int"_s);
+                return;
+            }
+        } while ((ancestorClass = ancestorClass->parentClass));
```

Comment:
That’s great, thank you! I will remove the `NumberObject`, `StringObject` and `BooleanObject` checks. Not sure I want to change this into a while loop, though. It would be elegant as a for loop, but that would add one unnecessary null check. For now I think I will leave it as a do/while.

Reply:
Oh, that's right: the parentClass relationship is static in our code base.  A quick grep reveals that only `NumberPrototype`, `StringPrototype`, and `BooleanPrototype` has these 3 classInfo as their parentClass.  But all 3 of those also do not have `JSType`s of `ObjectType` nor `FinalObjectType`.  Hence, it makes that we don't need to check for these when iterating the parentClass chain.

That said, there is also no classInfo that has `BigIntObject::s_info` as its parentClass.  So, you don't really need a loop here.  You only need to check if the top level classInfo is `BigIntObject::s_info`.

Reply:
I wonder if `BigIntObject` will pass the `canPerformFastPropertyEnumerationForJSONStringify` check. If it won’t, then I don’t need the `BigIntObject` check at all!

Reply:
There are many other checks, and if any are guaranteed false for `BigIntObject` then we can remove the check. For example, if the prototype is guaranteed to not be the default object prototype, or if `canPerformFastPropertyEnumerationForJSONStringify` will return false.

Reply:
OK, big int wrapped values have the `BigIntPrototype`, which is not the standard object prototype, so we don't need the `BigIntObject` check at all.

Reply:
BigIntObject does use the default object prototype.  AFAICT, it also doesn't set its own StructureFlags.  Hence, I think `canPerformFastPropertyEnumerationForJSONStringify` will return true.

Reply:
Oh you are correct.  I was checking the `CREATE_PROTOTYPE_FOR_SIMPLE_TYPE` macro, and conflated the prototype for `m_bigIntPrototype`'s base prototype (which is the default object prototype) for `m_bigIntPrototype` itself.  So yes, the `BigIntObject` check is not needed.